core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +669,824 @@ impl CheckedAdd for Fixed64 {
+					let dividend =
+						Double::from(self_norm.get(j + n))
+						* B
+						+ Double::from(self_norm.get(j + n - 1));
+					let divisor = other_norm.get(n - 1);
+					div_single(dividend, divisor.into())
+				};
+
+				// D3.1 test qhat
+				// replace qhat and rhat with RefCells. This helps share state with the closure
+				let qhat = RefCell::new(qhat);
+				let rhat = RefCell::new(rhat as Double);
+
+				let test = || {
+					// decrease qhat if it is bigger than the base (B)
+					let qhat_local = *qhat.borrow();
+					let rhat_local = *rhat.borrow();
+					let predicate_1 = qhat_local >= B;
+					let predicate_2 = {
+						let lhs = qhat_local * other_norm.get(n - 2) as Double;
+						let rhs = B * rhat_local + self_norm.get(j + n - 2) as Double;
+						lhs > rhs
+					};
+					if predicate_1 || predicate_2 {
+						*qhat.borrow_mut() -= 1;
+						*rhat.borrow_mut() += other_norm.get(n - 1) as Double;
+						true
+					} else {
+						false
+					}
+				};
+
+				test();
+				while (*rhat.borrow() as Double) < B {
+					if !test() { break; }
+				}
+
+				let qhat = qhat.into_inner();
+				// we don't need rhat anymore. just let it go out of scope when it does.
+
+				// step D4
+				let lhs = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+				let rhs = other_norm.clone().mul(Self::from(qhat));
+
+				let maybe_sub = lhs.sub(rhs);
+				let mut negative = false;
+				let sub = match maybe_sub {
+					Ok(t) => t,
+					Err(t) => { negative = true; t }
+				};
+				(j..=j+n).for_each(|d| { self_norm.set(d, sub.get(d - j)); });
+
+				// step D5
+				// PROOF: the `test()` specifically decreases qhat until it is below `B`. conversion
+				// is safe.
+				q.set(j, qhat as Single);
+
+				// step D6: add back if negative happened.
+				if negative {
+					q.set(j, q.get(j) - 1);
+					let u = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+					let r = other_norm.clone().add(u);
+					(j..=j+n).rev().for_each(|d| { self_norm.set(d, r.get(d - j)); })
+				}
+			}
+
+			// if requested, calculate remainder.
+			if rem {
+				// undo the normalization.
+				if normalizer_bits > 0 {
+					let s = SHIFT as u32;
+					let nb = normalizer_bits;
+					for d in 0..n-1 {
+						let v = self_norm.get(d) >> nb
+							| self_norm.get(d + 1).overflowing_shl(s - nb).0;
+						r.set(d, v);
+					}
+					r.set(n - 1, self_norm.get(n - 1) >> normalizer_bits);
+				} else {
+					r = self_norm;
+				}
+			}
+
+			Some((q, r))
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl rstd::fmt::Debug for Number{
+		fn fmt(&self, f: &mut rstd::fmt::Formatter<'_>) -> rstd::fmt::Result {
+			write!(
+				f,
+				"Number {{ {:?} ({:?})}}",
+				self.digits,
+				u128::try_from(self.clone()).unwrap_or_else(|_| 0),
+			)
+		}
+	}
+
+	impl PartialEq for Number {
+		fn eq(&self, other: &Self) -> bool {
+			// sadly, we have to reallocate here as strip mutably uses self.
```

Comment:
Wouldn't it be faster to just do a custom comparison ignoring the leading zero limbs than to reallocate unnecessarily? Or find the index of the first non-zero limbs and do a slice comparison.

Reply:
Yeah we could have the same approach but instead of `lstrip()`-ing we take slices and instead of `resize()`-ing we check if one of the slices has a bigger size. In that case we can early-return. makes sense?